lib/openstack_query/docs/user_docs/API.md

```diff
@@ -0,0 +1,612 @@
+- unique values found for a property specified in `group_by`
+- a set of pre-defined groups which define how to group results based on their value of the property specified in `group_by`
+
+```python
+def group_by(group_by: PropEnum,
+             group_ranges: Optional[Dict[str, List[PropValue]]] = None,
+             include_ungrouped_results: bool = False)
+```
+
+Public method used to configure how to group results.
+
+**Arguments**:
+
+- `group_by`: a property enum representing the property you want to group by
+    - e.g. If you're querying servers with `ServerQuery` - appropriate enum is `ServerProperties`
+
+- `group_ranges`: (optional) a dictionary of group mappings
+  - the keys are unique group names
+  - the values are a list of values that `group_by` property could have to be included in that group
+- `include_ungrouped_results`: (optional) flag that if true - will include an "ungrouped" group (Default is `False`)
+  - ungrouped group will contain all results that have a `group_by` property value that does not match any of the
+  groups set in `group_ranges`
+
+**Note**: You can group by properties you haven't 'selected' for using `select()`
+
+**Examples**
+Running an `ANY_IN` command and grouping results
+
+```python
+
+from openstack_query import ServerQuery
+from enums.query.props.server_properties import ServerProperties
+from enum.query.query_presets import QueryPresetsGeneric
+
+# create a ServerQuery
+query = ServerQuery()
+query.select(ServerProperties.SERVER_ID, ServerProperties.SERVER_NAME)
+
+# setup filter server_status = ERROR
+query.where(
+    preset=QueryPresetsGeneric.ANY_IN,
+    prop=ServerProperties.SERVER_STATUS,
+    values=["ERROR", "SHUTOFF"]
+)
+query.run("openstack-domain", as_admin=True, all_projects=True)
+query.group_by(ServerProperties.SERVER_STATUS)
+
+# holds a dictionary - where keys are unique values of "server_status" in results
+# in this case - ERROR and SHUTOFF since we queried for them
+x = query.to_objects()
+```
+
+```commandline
+> {"ERROR": [
+        {"server_id": 1, "server_name": "foo"},
+        {"server_id": 2, "server_name": "bar"},
+   ],
+   "SHUTOFF": [
+        {"server_id": 3, "server_name": "biz"},
+        {"server_id": 4, "server_name": "baz"},
+   ]
+```
+
+#
+### run
+
+`run()` will run the query. `run()` will apply all predefined conditioned set by `where` calls
+
+```python
+def run(cloud_account: Union[str, CloudDomains],
+        from_subset: Optional[List[OpenstackResourceObj]] = None,
+        **kwargs)
+```
+
+**Arguments**:
+
+- `cloud_account`: A String or a CloudDomains Enum representing the clouds configuration to use
+  - this should be the domain set in the `clouds.yaml` file located in `.config/openstack/clouds.yaml`
+
+- `from_subset`: (optional) a subset of openstack resources to run query on instead of querying for them using openstacksdk
+  - **NOTE:** this is going to be deprecated soon - look at chaining for a better way to do this
```

Comment:
Could we link the chaining docs here?

Reply:
I'm a bit apprehensive of linking developer docs here - I don't want to confuse the average user

What this should say is look at using `then` or `append_from` instead of `from_subset` - I'll change the wording